game-core/src/main/java/games/strategy/triplea/ai/tree/StepUnits.java

```diff
@@ -0,0 +1,446 @@
+  }
+
+  List<Unit> getAliveOrWaitingToDieEnemy() {
+    return filterUnits(this.enemyUnits, this.enemyBits, this.enemyUnits.size());
+  }
+
+  List<Unit> getAliveEnemy() {
+    return filterUnits(this.enemyUnits, this.enemyWaitingToDieBits, this.enemyUnits.size());
+  }
+
+  private List<Unit> filterUnits(final List<Unit> units, final BitSet bitSet, final int count) {
+    final List<Unit> filteredUnits = new ArrayList<>();
+    int pos = bitSet.nextClearBit(0);
+    while (pos < count) {
+      filteredUnits.add(units.get(pos));
+      pos = bitSet.nextClearBit(pos + 1);
+    }
+    return filteredUnits;
+  }
+
+  List<Unit> addFriendlyMultiHitTargets(final List<Unit> targets) {
+    final List<Unit> newTargets;
+    if (friendlyHitPoints.isEmpty()) {
+      newTargets = targets;
+    } else {
+      newTargets = new ArrayList<>(targets);
+      for (final Map.Entry<Unit, Integer> entry : friendlyHitPoints.entrySet()) {
+        if (!targets.contains(entry.getKey())) {
+          continue;
+        }
+        for (int i = 0; i < entry.getValue() - 1; i++) {
+          newTargets.add(0, entry.getKey());
+        }
+      }
+    }
+    return newTargets;
+  }
+
+  boolean noMoreFriendlies() {
+    return countOfFriendlyDamagedOrDead() == friendlyUnits.size() && retreatedFriendly.isEmpty();
+  }
+
+  boolean noMoreEnemies() {
+    return countOfEnemyDamagedOrDead() == enemyUnits.size() && retreatedEnemy.isEmpty();
+  }
+
+  int countOfFriendliesNotDamagedOrDead() {
+    return friendlyUnits.size() - countOfFriendlyDamagedOrDead();
+  }
+
+  int countOfFriendliesNotDead() {
+    return friendlyUnits.size() - countOfFriendlyDead();
+  }
+
+  int countOfEnemiesNotDamagedOrDead() {
+    return enemyUnits.size() - countOfEnemyDamagedOrDead();
+  }
+
+  int countOfEnemiesNotDead() {
+    return enemyUnits.size() - countOfEnemyDead();
+  }
+
+  int countOfFriendlyDamagedOrDead() {
+    return friendlyWaitingToDieBits.cardinality();
+  }
+
+  int countOfFriendlyDead() {
+    return friendlyBits.cardinality();
+  }
+
+  int countOfEnemyDamagedOrDead() {
+    return enemyWaitingToDieBits.cardinality();
+  }
+
+  int countOfEnemyDead() {
+    return enemyBits.cardinality();
+  }
+
+  int countOfFriendlyHitPoints() {
+    return friendlyHitPoints.values().stream().reduce(0, Integer::sum);
+  }
+
+  /** Store whether the units were alive */
+  void recordChances() {
+    for (final Unit unit : getAliveFriendly()) {
+      friendlyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : retreatedFriendly) {
+      friendlyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : getAliveEnemy()) {
+      enemyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : retreatedEnemy) {
+      enemyUnitsChances.put(unit, 1.0);
+    }
+  }
+
+  void updateUnitChances(final StepUnits child, final double probability) {
+    for (final Map.Entry<Unit, Double> entry : child.enemyUnitsChances.entrySet()) {
+      friendlyUnitsChances.compute(
```

Comment:
This case should never really occur if you ask me, Double should never be null, so I'd rather throw an exception here

Reply:
It can occur and does fairly often.  It happens when two child branches have different units hit.   This is merging the branch results so one branch will have a unit that has a probability for a unit and the other won't.  enemyUnitsChances/friendlyUnitsChances only has entries if there is a chance the unit might survive.

Reply:
Ah, nevermind.
I thought you were updating the same map you were reading from, which isn't the case 👍